reader.go

```diff
@@ -30,33 +30,140 @@ func (o *ReaderOptions) norm(name string) {
 
 // Reader reads data from a remote feed.
 type Reader struct {
-	remote *bfs.Object
-	opt    ReaderOptions
-	ctx    context.Context
-	num    int
+	ctx context.Context
+	opt ReaderOptions
 
-	br io.ReadCloser // bfs reader
-	cr io.ReadCloser // compression reader
-	fd FormatDecoder
+	remotes  []*bfs.Object
+	cur      *streamReader
+	pos, num int
 }
 
 // NewReader inits a new reader.
 func NewReader(ctx context.Context, remote *bfs.Object, opt *ReaderOptions) (*Reader, error) {
+	return MultiReader(ctx, []*bfs.Object{remote}, opt), nil
+}
+
+// MultiReader inits a new reader for multiple remotes.  Remotes are read sequentially as if concatenated.
+// Once all remotes are fully read, Read will return EOF.
+func MultiReader(ctx context.Context, remotes []*bfs.Object, opt *ReaderOptions) *Reader {
 	var o ReaderOptions
 	if opt != nil {
 		o = *opt
 	}
-	o.norm(remote.Name())
+	if len(remotes) > 0 {
+		o.norm(remotes[0].Name())
+	}
 
 	return &Reader{
-		remote: remote,
-		opt:    o,
-		ctx:    ctx,
-	}, nil
+		remotes: remotes,
+		opt:     o,
+		ctx:     ctx,
+	}
 }
 
 // Read reads raw bytes from the feed.
 func (r *Reader) Read(p []byte) (int, error) {
+	if !r.ensureCurrent() {
+		return 0, io.EOF
+	}
+
+	n, err := r.cur.Read(p)
+	if err == io.EOF {
+		// close and remove current reader
+		if err := r.cur.Close(); err != nil {
+			return n, err
+		}
+		r.cur = nil
+
+		// increment position and check if any more remotes
+		if r.pos++; r.pos < len(r.remotes) {
+			return r.Read(p) // start reading from next remote
```

Comment:
this was the one bit I was unsure of...  if you would ever end up with unread data from end of file if underlying reader returns `n, io.EOF` but from what I cans see in docs for all Readers i've seen it is always `n, nil` or `0, io.EOF` or `n, io.ErrUnexpectedEOF` eg. https://git.sr.ht/~tsileo/blobstash/tree/b178995e346d/pkg/filetree/reader/filereader/filereader.go#L312

Reply:
something to double-check, @mxmCherry pls

Reply:
have changed streamWriter to always read all data before returning EOF

Reply:
I have no idea about the internals - if it is returned with zero `n`, or partial read + EOF. I think this can depend on underlying reader implementation.

Just handle this in code, smth like:

```go
return r.Read(p[n:])
```

Double-check if it needs boundary checks (`n < len(p)`), don't remember how it behaves then, maybe even before getting into this branch (before opening next remote).

Reply:
what do you mean by the last part?  do you mean that reader might extend outside the size of the buffer you pass it?

Reply:
atually i think the simplest way is just to always return n, nil if EOF and there are more remotes.